examples/StandardFirmataWiFi/StandardFirmataWiFi.ino

```diff
@@ -902,37 +896,35 @@ void setup()
 #endif
 
   /*
-   * Configure WiFi security
+   * Configure WiFi security and initiate WiFi connection
    */
 #if defined(WIFI_WEP_SECURITY)
-  while (wifiStatus != WL_CONNECTED) {
     DEBUG_PRINT( "Attempting to connect to WEP SSID: " );
     DEBUG_PRINTLN(ssid);
-    wifiStatus = stream.begin( ssid, wep_index, wep_key, SERVER_PORT );
-    delay(5000); // TODO - determine minimum delay
-    if (++wifiConnectionAttemptCounter > WIFI_MAX_CONN_ATTEMPTS) break;
-  }
-
+  stream.begin(ssid, wep_index, wep_key);
 #elif defined(WIFI_WPA_SECURITY)
-  while (wifiStatus != WL_CONNECTED) {
     DEBUG_PRINT( "Attempting to connect to WPA SSID: " );
     DEBUG_PRINTLN(ssid);
-    wifiStatus = stream.begin(ssid, wpa_passphrase, SERVER_PORT);
-    delay(5000); // TODO - determine minimum delay
-    if (++wifiConnectionAttemptCounter > WIFI_MAX_CONN_ATTEMPTS) break;
-  }
-
+  stream.begin(ssid, wpa_passphrase);
 #else                          //OPEN network
-  while (wifiStatus != WL_CONNECTED) {
     DEBUG_PRINTLN( "Attempting to connect to open SSID: " );
     DEBUG_PRINTLN(ssid);
-    wifiStatus = stream.begin(ssid, SERVER_PORT);
-    delay(5000); // TODO - determine minimum delay
-    if (++wifiConnectionAttemptCounter > WIFI_MAX_CONN_ATTEMPTS) break;
-  }
+  stream.begin(ssid);
 #endif //defined(WIFI_WEP_SECURITY)
-
   DEBUG_PRINTLN( "WiFi setup done" );
+
+  /*
+   * Wait for TCP connection to be established
+   */
+  while (!streamConnected && ++connectionAttempts <= MAX_CONN_ATTEMPTS) {
+    delay(500);
+    streamConnected = stream.maintain();
+  }
+  if (streamConnected) {  
+    DEBUG_PRINTLN( "TCP connection established" );
+  } else {
+    DEBUG_PRINTLN( "failed to establish TCP connection" );    
```

Comment:
I didn't have a looked at the debug output so far.

Waiting until connected is a drawback if you add active logic that runs independently of the host. That is one reason why I wanted to introduce a [hostConnectedCallback](https://github.com/firmata/arduino/issues/280#issuecomment-204777188). 

Regarding the timeout, I reduced it by about 30% because an ESP8266 with static IP typically connects in less than 1.5 seconds. Raising the timeout should help somewhat.